### DISCUSSION TITLE
Proposal: Site-wide static data

### DISCUSSION THREAD

**Discussion by lostfictions** [vercel/next.js#14949] [Ideas]
# Feature request

## The problem

With `getStaticProps`, Next now has a great story for per-route build-time/SSG data, which can be generated by running arbitrary logic, accessing the filesystem or a database, etc. This data is calculated for each route and the client it loads it the first time they hit that route. No effort is made to chunk or deduplicate data across routes; if you include a chunk of data across several routes, it'll be redownloaded multiple times on the client as they navigate these routes.

As a concrete example, say you have a collection of markdown files you display on your site, and you want to build a table of contents for it that's shown on every page (or on most pages). To do this, you gather all the markdown files, read the frontmatter from each (hitting the filesystem) to grab its title and other relevant metadata, and create an object that contains each file's slug/route, title, etc.

You want to perform this work at build-time (and ideally in development it should be regenerated on reloads to reflect changes you might make to the markdown files). This is currently easy to achieve with `getStaticProps`, and you could factor out the TOC generation logic to a separate method that each route's `getStaticProps` invokes. But due to the lack of deduplication, this results in an extra, identical chunk of data being loaded by the client for every route. For small amounts of data, this overhead may be tolerable, but for larger payloads this cost could quickly become unacceptable.

Alternately, you could run a script at build-time to generate the file and include it as an npm script, eg. `"build": "node generate-toc.js && next build"`. The generated file could then simply be used wherever needed via an import, eg. `import TOC from '~/generated/toc.json'`, and if it's imported multiple times Webpack is smart enough to chunk it out so that it's not duplicated. But you'd be stuck having to rerun the generate-toc script manually in development on every change, or you'd have to integrate a watcher library like `chokidar` to detect changes, or otherwise add a lot of complexity to handle development cases. And you can't use Babel or TypeScript by default in this scenario without adding even more complexity.

Another alternative is to use something like [`babel-plugin-preval`](https://github.com/kentcdodds/babel-plugin-preval) or [`val-loader`](https://github.com/webpack-contrib/val-loader) to perform these precalculations. Both of them solve the issue of regeneration in development, but have their own idiosyncrasies, and require changing your Babel or Webpack configuration, respectively. And neither of them lets you use Babel or TypeScript for your own code by default either.

`val-loader` is what I most recently opted to use to solve this issue, but it would be great if Next supported site-wide static data out of the box. [`react-static`](https://github.com/react-static/react-static/) provides this feature via a configuration parameter called [`getSiteData`](https://github.com/react-static/react-static/blob/master/docs/config.md#getsitedata) and a corresponding [`useSiteData` hook](https://github.com/react-static/react-static/blob/master/docs/api.md#usesitedata), but I think Next could get away with something simpler.

## Describe the solution you'd like

I propose to allow declaring `getStaticProps` in `_app.js`. Any props generated this way would be handed to a custom `App` separately from the page props, allowing developers to hand them to a top-level React context provider, pass them directly to components, or whatever else they choose. These `getStaticProps` would only be computed once for the app, not for every route. Developers can use this in scenarios where a `NEXT_PUBLIC_` env var doen't suffice, or where code is otherwise required to create the required data.

This seems like the solution that's the least intrusive, introduces the fewest new concepts, doesn't expand the API surface, and is idiomatic to how Next currently works.

## Describe alternatives you've considered

Next could add an API similar to `react-static`, allowing a `getSiteData` parameter in `next.config.js` and some corresponding way to retrieve it in code (perhaps a hook, or some kind of `next/sitedata` façade package that allows a direct import. Conceptually this seems more complicated than the above solution.

**Comment by kristojorg** [comment]
I need this as well. I have an app with thousands of static pages, all of which require some site-wide data fetched from a (non-next) backend API. This API is rate limited. If I try to build my site now, the generation of every page will separately call the backend API, resulting in thousands of requests one after another. 

I need to be able to fetch this data once at static generation time.

**Comment by Vadorequest** [comment]
I also need this, and many other people too:

- https://github.com/vercel/next.js/discussions/13701
- https://github.com/vercel/next.js/discussions/18550
- https://github.com/vercel/next.js/discussions/14949

@Timer @timneutkens A few months ago, someone from Next/Vercel team told me it was a WIP, what's the status of this now? Is it intended to be built-in?

What I did to mitigate this issue was to build a server memoization, but it's not as efficient as pre-fetching the data once and make them available to all pages. See https://github.com/vercel/next.js/discussions/13765

I'd love a built-in solution for this, I'm also considering using val-loader, but I'd prefer not build a homemade solution and wait for an official one, if that's still on your roadmap.

**Reply by la55u** [reply]
does val-loader support the static regeneration feature of Nextjs?

**Comment by kristojorg** [comment]
I would also like to point out that there is a use-case for this that isn't just fetching one set of static data for the whole app. The more general case is where certain routes need shared static data which only needs to be fetched once. For example I might have `/[libraryId]/` where there might be 1000 pages beneath that all require the `libraryData`, but there are also 1000 possible libraries. I don't want to fetch the library data for every page individually, but I also don't want to fetch all 1000 on app startup. Having a way to pass static props to child routes (and adding this API to the index page) would solve both of these problems.

My proposal for that API here: #18052

**Comment by lostfictions** [comment]
For what it's worth, it does sound like this was already proposed and confirmed as in the works (or at least in the backlog) in #10949 -- I somehow missed that earlier discussion when creating this issue. (Unfortunate that discussions aren't searcheable alongside issues, nor are they sortable by reactions.)

**Reply by Vadorequest** [reply]
What issue are you talking about?

I couldn't find where it was mentioned that it was under work.

**Comment by ricokahler** [comment]
I just posted in this thread https://github.com/vercel/next.js/discussions/10949#discussioncomment-326570 and re-sending it along here.

tl;dr I have two libs around this problem that might interest y'all: [`next-plugin-preval`](https://github.com/ricokahler/next-plugin-preval) and [`next-data-hooks`](https://github.com/ricokahler/next-data-hooks) neither of them are slam dunks but it may help for the time being.
